In [1]:
import os
import pandas as pd

In [224]:
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])
df = school_data_complete
df.rename(columns = {'type':'schooltype'}, inplace = True)
#df.head(3)

In [29]:
totalschools = len(df["school_name"].unique())
#totalschools

In [ ]:
def place_comma(number): 
    return ("{:,}".format(number)) 

In [185]:
totalstudents = place_comma(df["student_name"].count())
#totalstudents

In [215]:
totalbudget = school_data["budget"].sum()
totalbudget = "$" + place_comma(totalbudget)
#totalbudget

In [176]:
averagemathscore = format(df["math_score"].sum() / totalstudents, '.2f')
#averagemathscore

In [175]:
averagereadingscore = format(df["reading_score"].sum() / totalstudents, '.2f')
#averagereadingscore

In [174]:
mathpass = df[df["math_score"] >= 70]
percentmathpass = format((len(mathpass) / totalstudents) * 100, '.2f')
#percentmathpass

In [172]:
readpass = df[df["reading_score"] >= 70]
percentreadpass = format((len(readpass) / totalstudents) * 100, '.2f')
#percentreadpass

In [173]:
overallpassingrate = format((averagemathscore + averagereadingscore) / 2, '.2f')
#overallpassingrate

In [226]:
data = {'Total Schools':[totalschools], 'Total Students': [totalstudents], 'Total Budget': [totalbudget], 
        'Average Math Score':[averagemathscore], 'Average Reading Score':[averagereadingscore], 
        '% Passing Math':[percentmathpass], '% Passing Reading': [percentreadpass], 
        '% Overall Passing Rate': [overallpassingrate]} 

newdf = pd.DataFrame(data) 

# District Summary

In [227]:
newdf

Total Schools Total Students Total Budget Average Math Score  \
0             15         39,170  $24,649,428              78.99   

  Average Reading Score % Passing Math % Passing Reading  \
0                 81.88          74.98             85.81   

  % Overall Passing Rate  
0                  80.43

In [231]:
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])
df = school_data_complete.set_index("school_name").sort_index()
df.rename(columns = {'type':'School Type'}, inplace = True)
df.drop(["School ID", "Student ID"], axis=1, inplace=True)
#df.head(3)

In [57]:
sn = df["School Type"]
sndf = pd.DataFrame(sn).reset_index()
grouped = sndf.drop_duplicates(subset = ["school_name"], keep = 'first')
school_type = grouped.set_index("school_name")
#school_type

In [14]:
mask1 = df["reading_score"] >= 70 
mask2 = df["math_score"] >= 70

pass_table = df[mask1]
pass_table2 = df[mask2]

In [218]:
read_pass = pass_table.pivot_table(values="student_name", index="school_name", aggfunc='count')
#read_pass

In [43]:
math_pass = pass_table2.pivot_table(values="student_name", index="school_name", aggfunc='count')
#math_pass

In [230]:
school_summary = pd.DataFrame(df.pivot_table(index="school_name"))
school_summary.rename(columns = {'budget':'Total School Budget', 'math_score':'Average Math Score',
                                 'reading_score': 'Average Reading Score', 'size': 'Total Students'}, inplace = True)
pd.options.display.float_format = '{:.2f}'.format
#school_summary

In [191]:
per_student_budget = pd.DataFrame({"Per Student Budget": school_summary["Total School Budget"]/school_summary["Total Students"]})
#per_student_budget

In [46]:
percent_math_pass = pd.DataFrame({"% Passing Math": (math_pass["student_name"] / school_summary["Total Students"]) * 100})
#percent_math_pass

In [47]:
percent_read_pass = pd.DataFrame({"% Passing Reading":(read_pass["student_name"] / school_summary["Total Students"]) * 100})
#percent_read_pass

In [48]:
overall_passing = pd.DataFrame({"% Overall Passing Rate": (percent_read_pass["% Passing Reading"] + percent_math_pass["% Passing Math"]) / 2})
#overall_passing

In [228]:
merged_data = pd.concat([school_summary,school_type, per_student_budget,percent_math_pass,percent_read_pass,overall_passing], axis=1)
del merged_data.index.name
merged_data = merged_data[['School Type', 'Total Students', 'Total School Budget', 'Per Student Budget', 'Average Math Score',
                            'Average Reading Score', '% Passing Math', '% Passing Reading', '% Overall Passing Rate']]

# School Summary

In [229]:
merged_data

School Type  Total Students  Total School Budget  \
Bailey High School       District            4976              3124928   
Cabrera High School       Charter            1858              1081356   
Figueroa High School     District            2949              1884411   
Ford High School         District            2739              1763916   
Griffin High School       Charter            1468               917500   
Hernandez High School    District            4635              3022020   
Holden High School        Charter             427               248087   
Huang High School        District            2917              1910635   
Johnson High School      District            4761              3094650   
Pena High School          Charter             962               585858   
Rodriguez High School    District            3999              2547363   
Shelton High School       Charter            1761              1056600   
Thomas High School        Charter            1635              1043130   
Wilson High School        Charter            2283              1319574   
Wright High School        Charter            1800              1049400   

                       Per Student Budget  Average Math Score  \
Bailey High School                 628.00               77.05   
Cabrera High School                582.00               83.06   
Figueroa High School               639.00               76.71   
Ford High School                   644.00               77.10   
Griffin High School                625.00               83.35   
Hernandez High School              652.00               77.29   
Holden High School                 581.00               83.80   
Huang High School                  655.00               76.63   
Johnson High School                650.00               77.07   
Pena High School                   609.00               83.84   
Rodriguez High School              637.00               76.84   
Shelton High School                600.00               83.36   
Thomas High School                 638.00               83.42   
Wilson High School                 578.00               83.27   
Wright High School                 583.00               83.68   

                       Average Reading Score  % Passing Math  \
Bailey High School                     81.03           66.68   
Cabrera High School                    83.98           94.13   
Figueroa High School                   81.16           65.99   
Ford High School                       80.75           68.31   
Griffin High School                    83.82           93.39   
Hernandez High School                  80.93           66.75   
Holden High School                     83.81           92.51   
Huang High School                      81.18           65.68   
Johnson High School                    80.97           66.06   
Pena High School                       84.04           94.59   
Rodriguez High School                  80.74           66.37   
Shelton High School                    83.73           93.87   
Thomas High School                     83.85           93.27   
Wilson High School                     83.99           93.87   
Wright High School                     83.95           93.33   

                       % Passing Reading  % Overall Passing Rate  
Bailey High School                 81.93                   74.31  
Cabrera High School                97.04                   95.59  
Figueroa High School               80.74                   73.36  
Ford High School                   79.30                   73.80  
Griffin High School                97.14                   95.27  
Hernandez High School              80.86                   73.81  
Holden High School                 96.25                   94.38  
Huang High School                  81.32                   73.50  
Johnson High School                81.22                   73.64  
Pena High School                   95.95                   95.27  
Rodriguez High School             

## Top Performing Schools (By Passing Rate)

In [194]:
top_performing_schools = merged_data.sort_values("% Overall Passing Rate", ascending = False)
top_performing_schools.head(5)

School Type  Total Students  Total School Budget  \
Cabrera High School     Charter            1858              1081356   
Thomas High School      Charter            1635              1043130   
Pena High School        Charter             962               585858   
Griffin High School     Charter            1468               917500   
Wilson High School      Charter            2283              1319574   

                     Per Student Budget  Average Math Score  \
Cabrera High School              582.00               83.06   
Thomas High School               638.00               83.42   
Pena High School                 609.00               83.84   
Griffin High School              625.00               83.35   
Wilson High School               578.00               83.27   

                     Average Reading Score  % Passing Math  % Passing Reading  \
Cabrera High School                  83.98           94.13              97.04   
Thomas High School                   83.85           93.27              97.31   
Pena High School                     84.04           94.59              95.95   
Griffin High School                  83.82           93.39              97.14   
Wilson High School                   83.99           93.87              96.54   

                     % Overall Passing Rate  
Cabrera High School                   95.59  
Thomas High School                    95.29  
Pena High School                      95.27  
Griffin High School                   95.27  
Wilson High School                    95.20

## Bottom Performing Schools (By Passing Rate)

In [195]:
bottom_performing_schools = merged_data.sort_values("% Overall Passing Rate")
bottom_performing_schools.head(5)

School Type  Total Students  Total School Budget  \
Rodriguez High School    District            3999              2547363   
Figueroa High School     District            2949              1884411   
Huang High School        District            2917              1910635   
Johnson High School      District            4761              3094650   
Ford High School         District            2739              1763916   

                       Per Student Budget  Average Math Score  \
Rodriguez High School              637.00               76.84   
Figueroa High School               639.00               76.71   
Huang High School                  655.00               76.63   
Johnson High School                650.00               77.07   
Ford High School                   644.00               77.10   

                       Average Reading Score  % Passing Math  \
Rodriguez High School                  80.74           66.37   
Figueroa High School                   81.16           65.99   
Huang High School                      81.18           65.68   
Johnson High School                    80.97           66.06   
Ford High School                       80.75           68.31   

                       % Passing Reading  % Overall Passing Rate  
Rodriguez High School              80.22                   73.29  
Figueroa High School               80.74                   73.36  
Huang High School                  81.32                   73.50  
Johnson High School                81.22                   73.64  
Ford High School                   79.30                   73.80

## Math Scores by Grade

In [120]:
score_by_grade = df
score_by_grade_df = pd.DataFrame(score_by_grade).reset_index().drop(["student_name", "gender", "School Type", "size", "budget"], axis=1)
#score_by_grade_df

In [232]:
math_score_table = score_by_grade_df.pivot_table(values="math_score", index=("school_name"), aggfunc='mean', columns = "grade")

del math_score_table.index.name
del math_score_table.columns.name

column_order = ['9th', '10th', '11th', '12th']
new_math_score_table = math_score_table.reindex(column_order, axis=1)

In [233]:
new_math_score_table

9th  10th  11th  12th
Bailey High School    77.08 77.00 77.52 76.49
Cabrera High School   83.09 83.15 82.77 83.28
Figueroa High School  76.40 76.54 76.88 77.15
Ford High School      77.36 77.67 76.92 76.18
Griffin High School   82.04 84.23 83.84 83.36
Hernandez High School 77.44 77.34 77.14 77.19
Holden High School    83.79 83.43 85.00 82.86
Huang High School     77.03 75.91 76.45 77.23
Johnson High School   77.19 76.69 77.49 76.86
Pena High School      83.63 83.37 84.33 84.12
Rodriguez High School 76.86 76.61 76.40 77.69
Shelton High School   83.42 82.92 83.38 83.78
Thomas High School    83.59 83.09 83.50 83.50
Wilson High School    83.09 83.72 83.20 83.04
Wright High School    83.26 84.01 83.84 83.64

## Reading Score by Grade

In [234]:
read_score_table = score_by_grade_df.pivot_table(values="reading_score", index=("school_name"), aggfunc='mean', columns = "grade")

del read_score_table.index.name
del read_score_table.columns.name

column_order = ['9th', '10th', '11th', '12th']
new_read_score_table = read_score_table.reindex(column_order, axis=1)

In [235]:
new_read_score_table

9th  10th  11th  12th
Bailey High School    81.30 80.91 80.95 80.91
Cabrera High School   83.68 84.25 83.79 84.29
Figueroa High School  81.20 81.41 80.64 81.38
Ford High School      80.63 81.26 80.40 80.66
Griffin High School   83.37 83.71 84.29 84.01
Hernandez High School 80.87 80.66 81.40 80.86
Holden High School    83.68 83.32 83.82 84.70
Huang High School     81.29 81.51 81.42 80.31
Johnson High School   81.26 80.77 80.62 81.23
Pena High School      83.81 83.61 84.34 84.59
Rodriguez High School 80.99 80.63 80.86 80.38
Shelton High School   84.12 83.44 84.37 82.78
Thomas High School    83.73 84.25 83.59 83.83
Wilson High School    83.94 84.02 83.76 84.32
Wright High School    83.83 83.81 84.16 84.07

## Scores by School Spending

In [204]:
school_spend = merged_data.reset_index()
#school_spend

In [236]:
spending_bins = [0, 585, 615, 645, 675]
group_names = ["<$585", "$585-615", "$615-645", "$645-675"]

school_spend["Spending Ranges (Per Student)"] = pd.cut(school_spend["Per Student Budget"], spending_bins, labels=group_names)
school_spend_df = school_spend.drop(["index", "School Type", "Total Students", "Total School Budget", "Per Student Budget"], axis=1)

school_spend_df = school_spend_df.groupby("Spending Ranges (Per Student)")

In [237]:
school_spend_df.mean()

Average Math Score  Average Reading Score  \
Spending Ranges (Per Student)                                              
<$585                                       83.46                  83.93   
$585-615                                    83.60                  83.89   
$615-645                                    79.08                  81.89   
$645-675                                    77.00                  81.03   

                               % Passing Math  % Passing Reading  \
Spending Ranges (Per Student)                                      
<$585                                   93.46              96.61   
$585-615                                94.23              95.90   
$615-645                                75.67              86.11   
$645-675                                66.16              81.13   

                               % Overall Passing Rate  
Spending Ranges (Per Student)                          
<$585                                           95.04  
$585-615                                        95.07  
$615-645                                        80.89  
$645-675                                        73.65

## Scores by School Size

In [201]:
school_size = merged_data.reset_index()
#school_size

In [238]:
size_bins = [0, 1000, 2000, 5000]
group_names = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

school_size["School Size"] = pd.cut(school_size["Total Students"], size_bins, labels=group_names)
school_size_df = school_size.drop(["index", "School Type", "Total Students", "Total School Budget", "Per Student Budget"], axis=1)

school_size_df = school_size_df.groupby("School Size")

In [239]:
school_size_df.mean()

Average Math Score  Average Reading Score  % Passing Math  \
School Size                                                                     
Small (<1000)                    83.82                  83.93           93.55   
Medium (1000-2000)               83.37                  83.86           93.60   
Large (2000-5000)                77.75                  81.34           69.96   

                    % Passing Reading  % Overall Passing Rate  
School Size                                                    
Small (<1000)                   96.10                   94.82  
Medium (1000-2000)              96.79                   95.20  
Large (2000-5000)               82.77                   76.36

## Scores by School Type

In [240]:
school_type_table = merged_data.reset_index()
school_type_table_df = school_type_table.drop(["index", "Total Students", "Total School Budget", "Per Student Budget"], axis=1)
school_type_table_df = school_type_table_df.groupby("School Type")

In [241]:
school_type_table_df.mean()

Average Math Score  Average Reading Score  % Passing Math  \
School Type                                                              
Charter                   83.47                  83.90           93.62   
District                  76.96                  80.97           66.55   

             % Passing Reading  % Overall Passing Rate  
School Type                                             
Charter                  96.59                   95.10  
District                 80.80                   73.67

## Trends

1) Students at Charter Schools are better academic performers than those in District Schools.

2) District schools have a higher population which is most likely a factor in trend #1. When schools have a higher population,
   each classroom has a higher number of students which then leads to lack of sufficient support from one teacher to students who 
   need help or are falling behind.
   
3) The 5 bottom performing schools all have low average math scores (77 and below) with an average of only 66% passing rate. This is an 
   indication that the math program in all grade levels (9th - 12th) needs much improvement. Perhaps the schools need to hire better math    
   teachers. Or create a math tutoring after school program - the budget per student at these schools are about $645 compared to just $488 at 
   the top performing schools. It makes one wonder where that budget is focused on at the bottom performing schools.
